lib/Sema/SemaCast.cpp

```diff
@@ -2565,7 +2565,15 @@ void CastOperation::CheckCStyleCast() {
     unsigned TypeKind = 0;
     bool HasUncheckedType = DestType->hasUncheckedType(TypeKind);
     bool HasVariadicType = DestType->hasVariadicType();
-    if (HasUncheckedType || HasVariadicType) {
+    bool ConstructsNullPointer = false;
+    if (HasUncheckedType)
+      if (DestType->isVoidPointerType() &&
+          !SrcExpr.isInvalid()) {
+        const IntegerLiteral *Lit = dyn_cast<IntegerLiteral>(SrcExpr.get());
+        if (Lit && !Lit->getValue())
+         ConstructsNullPointer = true;
+      }
```

Comment:
Looks good :)

I have a question about this strategy.
Is only void type casting of zero `(void*)0` allowed?
non-void type casting of zero such as` (int*)0` or `(ptr<int>)0` is not allowed?